### ISSUE TITLE
SPIR-V conditional branch miscompilation

### ISSUE THREAD

**Issue by anon-apple** [IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT#318]
### Checklist [README]

- [X] Device is not undervolted nor overclocked
- [X] Device is using the [latest drivers](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/wiki/Intel-driver-updates-and-changelogs)
- [X] Application is not cracked, modded and use the latest patch

### Application [Required]

RTX Remix

### Processor / Processor Number [Required]

Intel Core i9-7900X

### Graphic Card [Required]

Intel Arc A770 16GB

### GPU Driver Version [Required]

- [X] 31.0.101.4369
- [ ] 31.0.101.4335
- [ ] 31.0.101.4314
- [ ] 31.0.101.4311
- [ ] 31.0.101.4257
- [ ] 31.0.101.4255
- [ ] 31.0.101.4148
- [ ] 31.0.101.4146
- [ ] 31.0.101.4125
- [ ] 31.0.101.4123
- [ ] 31.0.101.4091
- [ ] 31.0.101.4090
- [ ] 31.0.101.4034
- [ ] 31.0.101.4032
- [ ] 31.0.101.3975

### Rendering API [Required]

- [X] Vulkan
- [ ] OpenGL
- [ ] DirectX12
- [ ] DirectX11
- [ ] DirectX10
- [ ] DirectX9
- [ ] Not applicable

### Windows Build Number [Required]

- [ ] Windows 11 22H2
- [ ] Windows 11 21H2
- [ ] Windows 10 22H2
- [ ] Windows 10 21H2
- [X] Other (Please specify)

### Other Windows build number

Windows 10 20H2

### Intel System Support Utility report

[igcit_ssu.txt](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/files/11502102/igcit_ssu.txt)

### Description and steps to reproduce [Required]

While trying to get RTX Remix to work on Intel dedicated GPUs some sort of incorrect SPIR-V compilation behavior causes various rendering issues which will likely have to be worked around. I have isolated the problem out of the RTX Remix codebase into a minimal example as well to make this bug report simpler.

In short, the problem emerges due to code with this sort of structure:
```
Bar bar;
bar.b = false;

if (!bool(constants.f) && !bar.b) {
  return;
} else if (bool(constants.f) && !bar.b) {
  return;
}
```
`constants.f` is a unsigned integer push constant set to the value 1, and `bar.b` is a boolean struct member which is always set to false. With this known this code is expected to always return on one of these paths, but in reality neither return is hit and the code proceeds past this point for some reason. Oddly this only happens as well when generating optimized SPIR-V from GLSL, so a functionally similar SPIR-V generated without optimizations can be provided for comparison to help track the issue down further.

In the minimal example I constructed, I based it off of [this](https://github.com/Erkaman/vulkan_minimal_compute) minimal Vulkan open source project, modified so that it outputs an image with a solid color to indicate which path is being taken. A green image output indicates that one of the returns was taken (which is the expected behavior), red indicates they both were skipped.

The shader source, compiled buggy (optimized) and non-buggy (normal) SPIR-V are available here for examination:
[shaders.zip](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/files/11502179/shaders.zip)

The whole Vulkan project around this shader can be found here:
[vulkan_minimal_compute-master.zip](https://github.com/IGCIT/Intel-GPU-Community-Issue-Tracker-IGCIT/files/11502195/vulkan_minimal_compute-master.zip)

Building/running it is fairly simple, simply set up the project with CMake:
```
mkdir build
cd build
cmake ..
```
Then open the solution (if using Visual Studio), build the project, and run. Running will generate an `output.png` file with the color either red or green based on if the bug happened (it's red for me).

I should note it is possible that this is a result of the `-Os` flag in glslang generating bad SPIR-V as I haven't yet tested this exact minimal example on Nvidia GPUs yet, but since the code in RTX Remix this isolated example is based on works fine on Nvidia/AMD cards and does not on Intel I am more inclined to think it is a specific SPIR-V to ISA compilation issue in the Intel driver (though I will test this example on a Nvidia GPU here soon to confirm that this is the case).

**Comment by Arturo-Intel** [comment]
Hi @anon-apple 

I will replicate this issue in the lab, it looks similar to #159 which is already being worked by the drivers team

Keep you posted
--r2

**Comment by Arturo-Intel** [comment]
Hi @anon-apple 

I was able to reproduce this behavior with the -O and -Os shader compilation flags in Arc 770
I can confirm this is not happening on 3060RTX

I was hopping that the fix implemented for #159 would help us with this issue, but is seems not.
I will continue with the escalation process, hope we can have a fix soon.
-- r2

**Comment by Karen-Intel** [comment]
Hey all, this issue has been fixed w/driver v/4284. @IGCIT could you help us close this one? TY

Karen
